### DISCUSSION TITLE
Unfill-paragraph when converting commit messages to PR descriptions

### DISCUSSION THREAD

**Discussion by xmo-odoo** [community/community#21065] [Pull Requests] [answered]
Conventionally, commit messages should be fill-paragraph'd to 72 characters (except the first line which should be limited to 50).

However when Github picks that up to generate a garbage-looking PR message since GFM interprets line breaks as line breaks but double breaks as paragraphs.

This is made worse by the visual expansion and contraction of references e.g. a #-reference to an issue or PR gets expanded with the entire title of the corresponding object, but a 40 characters hash reference gets shortened to 7~8. Links and footnotes in markdown-style also get altered by the transition.

If github "unfilled" the paragraphs of text (removed the linebreaks but kept the paragraph breaks) it'd make PRs much better looking out of the box for people who write & format extensive commit messages.

**Comment by brasic** [comment, marked as answer]
@xmo-odoo would you be interested in beta testing this feature?

Edit: [this is shipped](https://github.blog/changelog/2022-09-07-better-suggested-pull-request-description-from-commit-message/)!

**Reply by xmo-odoo** [reply]
@brasic sure, it will be rate limited to my creation of PRs but I'd be happy to test it.

**Reply by brasic** [reply]
@xmo-odoo I've enabled your account. Please feel free to leave any feedback here!

**Reply by xmo-odoo** [reply]
@brasic super happy with the feature so far, I haven't had the occasion to test it with code blocks yet (as I haven't had a reason to add a code block in a commit message in the two weeks since your message), and that's what I'm the most wary of, but it works stellar for regular paragraphs and lists and things like that.

**Reply by xmo-odoo** [reply]
Seems to have been deployed yesterday: https://github.blog/changelog/2022-09-07-better-suggested-pull-request-description-from-commit-message/

Not sure if there's something I can do to mark the issue as resolved or anything, anyway thanks @brasic

**Reply by brasic** [reply]
Indeed! Thanks for the report and for testing a fix!

**Reply by jstasiak** [reply]
@brasic is there a way to turn this feature off? It's actively breaking my PR descriptions by joining lines that should never be joined, it's really frustrating to have to fix them all the time by hand.

**Reply by brasic** [reply]
@jstasiak there is not. Can you share an example? If it’s sensitive feel free to email me at brasic@github.com.

**Reply by jstasiak** [reply]
An example commit message (based on real ones I've made recently):

```
Do something

Blah blah when I did [1] I didn't know that blah blah in [2] and then
blah blah blah so that blabh blabh blah. An illustration of why [1]
was blah blah blah:

Fixes: 1234567 ("Blah blah blah blaha blah blah (#1234)")
Some other thing: https://whatever/some/url/goes/here etc
Resolves: LLL-123 https://some/long/url/here and some other stuff
Fixes: 9876543 ("Other blah blah blah blah (#1111))
Amends: LLL-321

[1] 111111111111 ("xxxxxx xxxxxxxx xxxxx xxxxxx xxxxxxx xxxxxxx xxxxxxxx (#123)")
[2] 222222222222 ("xxxxxxxxxx xxxxnxxxn xxxxx xxxxxxx xxxxxx (#124)")
```

This is what it gets converted to in a PR description:

```
Blah blah when I did [1] I didn't know that blah blah in [2] and then blah blah blah so that blabh blabh blah. An illustration of why [1] was blah blah blah:

Fixes: 1234567 ("Blah blah blah blaha blah blah (#1234)") Some other thing: https://whatever/some/url/goes/here etc Resolves: LLL-123 https://some/long/url/here and some other stuff Fixes: 9876543 ("Other blah blah blah blah (#1111)) Amends: LLL-321

[1] 111111111111 ("xxxxxx xxxxxxxx xxxxx xxxxxx xxxxxxx xxxxxxx xxxxxxxx (#123)") [2] 222222222222 ("xxxxxxxxxx xxxxnxxxn xxxxx xxxxxxx xxxxxx (#124)")
```

Note how mangled is it. Additionally this mangled version then is set as the default commit message when squash merging.

**Reply by jstasiak** [reply]
@brasic Can a feature switch be introduced for this or the feature be reverted? It continues to be a problem and it produces invalid, mangled PR descriptions.

**Reply by brasic** [reply]
@jstasiak taking a look.

**Reply by jstasiak** [reply]
Thank you @brasic, any luck so far? This continues to be a significant problem.

**Reply by jstasiak** [reply]
Hey @brasic, have you found anything interesting?

**Reply by jstasiak** [reply]
@brasic Do we know more about this issue now? It's still present and it continues to break PR descriptions

**Reply by brasic** [reply]
Hey @jstasiak, I've pushed an improvement to the unwrapping logic. Let me know if you are still seeing issues.

**Reply by jstasiak** [reply]
Hey @brasic thank you for working on this. The issue is resolved in many cases but I still see it reappear every now and then. The latest one I just got, this commit message

```
Do something

1. Grab the aaaaaaaaa aaaaa database overview file[1][2][3].
   Size: 12345678
   SHA256 sum: 1231231231231231231231231231231231231231231231231231231231231231
2. Extract the "asdasdasdasd" sheet.
```

gets converted to this PR description (the title is `Do something` as expected)

```
1. Grab the aaaaaaaaa aaaaa database overview file[1][2][3]. Size: 12345678 SHA256 sum: 1231231231231231231231231231231231231231231231231231231231231231
2. Extract the "asdasdasdasd" sheet.
```

**Reply by jstasiak** [reply]
@brasic another example of a broken PR description

Commit message:

```
Bla bla bla

The actual response type looks like this:

    interface MapiResponse<T = any> {
        /**
         * The response body, parsed as JSON.
         */
        body: T;
        /**
         * The raw response body.
         */
        rawBody: string;
        /**
         * The response's status code.
         */
        statusCode: number;
        /**
         * The parsed response headers.
         */
        headers: any;
        /**
         * The parsed response links
         */
        links: any;
        /**
         * The response's originating MapiRequest.
         */
        request: MapiRequest<T>;
        hasNextPage(): boolean;
        nextPage(): MapiRequest<T> | null;
    }
```

the PR description generated from the above:

```
The actual response type looks like this:

    interface MapiResponse<T = any> {
        /**
         * The response body, parsed as JSON. */ body: T; /**
         * The raw response body. */ rawBody: string; /** * The response's status code. */ statusCode: number; /** * The parsed response headers. */ headers: any; /** * The parsed response links */ links: any; /** * The response's originating MapiRequest. */ request: MapiRequest<T>; hasNextPage(): boolean; nextPage(): MapiRequest<T> | null; }
```

Needless to say it's a problem.

**Reply by jstasiak** [reply]
@brasic another random case

Commit message

```
Something something

As a bonus this allows us to remove some potentially unsafe type
assertions.
    
[1] https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String#string_coercion
[2] One exception would be if toString() did something other than we
    wanted, but this is not the case here.
```

PR description generated from the above:

```
As a bonus this allows us to remove some potentially unsafe type assertions.

[1] https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String#string_coercion [2] One exception would be if toString() did something other than we
    wanted, but this is not the case here.
```

This is really annoying.

Can this behavior be completely disabled or at least a switch introduced to disabled it for a given GH user?

**Reply by jstasiak** [reply]
@brasic The issues described above are still live. Is there a timeline for when this problem is resolved? It's a really disruptive bug that the PR descriptions are mangled like this.

**Reply by brasic** [reply]
I don't work at GitHub anymore.
